File: prob1.py
# Time Complexity :  O(2**n)
# Space Complexity :O(h)
# Passed on Leetcode: yes

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.result = []
        self.path = []
        self.backtrack(nums, self.path, self.result, 0)
        return self.result

    def backtrack(self, nums, path, result, idx):
        # base
        if idx == len(nums):
            result.append(copy.deepcopy(path))
            return

        # logic

        # not choose
        self.backtrack(nums, path, result, idx+1)

        # choose
        # action
        path.append(nums[idx])
        self.backtrack(nums, path, result, idx+1)

        # backtrack
        path.pop()

File: prob1.java
// Time Complexity : O(n^2)
// Space Complexity : O(n^2)
// Did this code successfully run on Leetcode : Yes


class Solution {
    static void subsequence(int i,int[] arr, List<Integer> lst, List<List<Integer>> res){
        if(res.contains(lst)) return;
        if(i==arr.length){
            res.add(new ArrayList<>(lst));
            return;
        }
        lst.add(arr[i]);
        subsequence(i+1, arr, lst, res);
        lst.remove(lst.size()-1);
        subsequence(i+1, arr, lst, res);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res=new ArrayList<>();
        subsequence(0, nums, new ArrayList<>(), res);
        return res;
    }
}